### PR TITLE
TSK-1967: make owner of Task updatable when Task is 'READY_F…OR_REVIEW'

### PR DESCRIPTION
**Pull request by norman-schmidt**
---
Release Notes:

```
**Improvement**
* TSK-1967: setting the owner of a Task now works when the TaskState is READY_FOR_REVIEW 

```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
https://sonarcloud.io/summary/new_code?id=norman-schmidt_taskana&branch=TSK-1967
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: common/taskana-common-data/src/main/resources/sql/sample-data/task.sql [hunk 1]
```diff
@@ -1,101 +1,102 @@
...
 
--- TASK TABLE          (ID                                        , EXTERNAL_ID                               , CREATED             , CLAIMED              , COMPLETED            , modified             , received             , planned              , due                  , name                      , creator                      , description                                , note                                     , priority, manual_priority, state       , classification_category  , classification_key, classification_id                         , workbasket_id                              , workbasket_key, domain    , business_process_id, parent_business_process_id, owner         , por_company , por_system , por_system_instance, por_type , por_value    , is_read, is_transferred, callback_info, callback_state, custom_attributes , custom1     , custom2     , custom3,   , custom4    , custom5    , custom6    , custom7    ,custom8     ,custom9     ,custom10    ,custom11,    ,custom12     ,custom13     ,custom14  ,custom15     ,custom16    , custom-int-1, custom-int-2, custom-int-3, custom-int-4, custom-int-5, custom-int-6, custom-int-7, custom-int-8
```
Comment (mustaphazorgati) [inline common/taskana-common-data/src/main/resources/sql/sample-data/task.sql:3]: What did you modify in this file?

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 2]
```diff
@@ -2013,11 +2013,13 @@ private void standardUpdateActions(TaskImpl oldTaskImpl, TaskImpl newTaskImpl)
       newTaskImpl1.setBusinessProcessId(oldTaskImpl.getBusinessProcessId());
     }
 
-    // owner can only be changed if task is in state ready
+    // owner can only be changed if task is either in state ready or ready_for_review
     boolean isOwnerChanged = !Objects.equals(newTaskImpl1.getOwner(), oldTaskImpl.getOwner());
-    if (isOwnerChanged && oldTaskImpl.getState() != TaskState.READY) {
+    if (isOwnerChanged
+        && oldTaskImpl.getState() != TaskState.READY
+        && oldTaskImpl.getState() != TaskState.READY_FOR_REVIEW) {
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:2020]:
```suggestion
        && !Set.of(TaskState.READY, TaskState.READY_FOR_REVIEW).contains(oldTaskImpl.getState())) {
```

Comment (mustaphazorgati) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:2020]: What do u think is more readable?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 3]
```diff
@@ -1534,6 +1534,34 @@ void should_UpdateTaskOwnerOfReadyTask() {
...
+        new HttpEntity<>(theTaskRepresentationModel, RestHelper.generateHeadersForUser("user-1-2"));
+
+    ResponseEntity<TaskRepresentationModel> responseUpdate =
+        TEMPLATE.exchange(url, HttpMethod.PUT, auth2, TASK_MODEL_TYPE);
+
+    assertThat(responseUpdate.getBody()).isNotNull();
+    TaskRepresentationModel theUpdatedTaskRepresentationModel = responseUpdate.getBody();
+    assertThat(theUpdatedTaskRepresentationModel.getState()).isEqualTo(TaskState.READY_FOR_REVIEW);
+    assertThat(theUpdatedTaskRepresentationModel.getOwner()).isEqualTo("dummyUser");
+  }
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1563]: Why are you testing this on REST level? I expected a test in taskana-core or taskana-core-test

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]
I am missing a fix for the bulk operation aswell :wink:

File: common/taskana-common-data/src/main/resources/sql/sample-data/task.sql (hunk 1 above)
Comment (norman-schmidt) [reply common/taskana-common-data/src/main/resources/sql/sample-data/task.sql:3]: I added a new Task (TKI:000000000000000000000000000000000104) with the state READY_FOR_REVIEW:

image

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java (hunk 2 above)
Comment (norman-schmidt) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:2020]: Also thought about that but thought I'd stick to && statement. I think mine might be more readable but yours is less boilerplate and kind of fancier so I will implement the suggestion.

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 3 above)
Comment (norman-schmidt) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1563]: I'll add a test there, but would leave this test as we are also testing it for state READY on REST level.

File: lib/taskana-core/src/test/java/acceptance/task/claim/SetOwnerAccTest.java [hunk 4]
```diff
@@ -50,6 +50,25 @@ void testSetOwnerAndSubsequentClaimSucceeds() throws Exception {
     assertThat(taskClaimed.getOwner()).isEqualTo("user-1-2");
   }
 
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void testSetOwnerOfReadyForReviewTaskAndSubsequentClaimSucceeds() throws Exception {
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/claim/SetOwnerAccTest.java:55]: this naming is not according to our guidelines :wink:

**Review by mustaphazorgati — DISMISSED** [review]
Where is the test for the bulk operation? :D

File: lib/taskana-core/src/test/java/acceptance/task/claim/SetOwnerAccTest.java (hunk 4 above)
Comment (norman-schmidt) [reply lib/taskana-core/src/test/java/acceptance/task/claim/SetOwnerAccTest.java:55]: I thought I stick to the naming of the tests in this class and keep the refactoring for a seperate ticket ^^

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**